### ISSUE TITLE
First protocol is used instead of anyprot in certain conditions

### ISSUE THREAD

**Issue by shadowlmd** [yrutschle/sslh#195]
In a condition when every probe returned PROBE_AGAIN and there were no more bytes read from the socket, first protocol is executed instead of "anyprot". To reproduce this bug it is enough to do something like this:
```
echo | nc 127.0.0.1 443
```
This is a corner case and does not have a serious impact on anything, at least I can't think of any real life example, but it can be confusing when one tests regexp probes, for example. I could not find a good way to fix this so here's another ugly hack just for the reference: https://github.com/shadowlmd/sslh/commit/5b1a3ea31db6361b66cb23885f381c41c8102c4c (sorry, this commit has some leftovers not related to this issue, but they are obvious).

**Comment by yrutschle** [comment]
On Mon, Jul 30, 2018 at 11:57:37PM +0000, Vasya Pupkin wrote:
> In a condition when every probe returned PROBE_AGAIN and there were no more bytes read from the socket, first protocol is executed instead of "anyprot".

Actually this would trigger a timeout and forward to
'timeout' protocol, or first protocol if not specified (as
specified in... a comment before the timeout_protocol()
function, and contrary to what's written in the man page).

I'll fix the man page...

**Comment by shadowlmd** [comment]
I guess a documented bug can be called a feature. But can you at least print something to log in case this happens so people like me will understand why incorrect protocol is triggered? :)

**Comment by yrutschle** [comment]
Really not a bug, timing out ("nothing is happening on that
socket, what do I do now?" (usually, connect to ssh)) and
anyprot ("There is stuff on that socket but I have no idea
what it is, what do I do now?" (usually, connect to some
obscure protocol only you use and for which no probe exists,
or can be written) are very, very different things.  In fact
that's how sslh worked before there were probes.

There is already a log at level LOG_INFO saying
"timed out, connect to <some protocol>".

**Comment by shadowlmd** [comment]
But there is something on the socket. Some bytes are there, they were just not enough for first probe to return something other than PROBE_AGAIN. So it is actually a bug. We received some data, say, 2 bytes, first probe (i.e. tls) returned PROBE_AGAIN (and because of #194 other probes did not have a chance to do probing), no more bytes came in and "timeout" (or first defined proto) was used. This can be really confusing when you try to figure out why the hell data like `\xFF\xFD` can trigger "tls" proto.

And no, this entry never appears in log, and in fact, timeout protocol never triggers as well. Please check. This is my config:
```
protocols:
(
     { name: "socks5";  host: "127.0.0.1"; port: "12345"; },
     { name: "openvpn"; host: "127.0.0.1"; port: "12346"; },
     { name: "tls";     host: "127.0.0.1"; port: "12347"; },
     { name: "anyprot"; host: "127.0.0.1"; port: "12348"; }
);
on-timeout: "tls";
```
I am running this:
```
echo -e '\r\r' | nc localhost 443
```
And getting this in log:
```
Aug  1 01:36:37 server sslh-fork[23335]: socks5:connection from 127.0.0.1:55996 to 127.0.0.1:443 forwarded from 127.0.0.1:52982 to 127.0.0.1:12345
```
According to code, this will always happen as long as any probe returns PROBE_AGAIN because you have this test:
```if (n > 0) {```
On line 391 of probe.c. So when 0 bytes were read during next `probe_client_protocol()` iteration, this code is being executed:
```
    if (verbose)
        fprintf(stderr,
                "all probes failed, connecting to first protocol: %s\n",
                protocols->description);

    /* If none worked, return the first one affected (that's completely
     * arbitrary) */
    cnx->proto = protocols;
    return PROBE_MATCH;
```

**Comment by shadowlmd** [comment]
I think I understand where our misunderstanding comes from. I am speaking about a situation where a client is sending few bytes and **closes connection immediately** after that. Only this scenario triggers the bug. If client is not closing connection, it is indeed forwarded to timeout protocol and that is correct. But if client closes connection, connection is forwarded to first defined protocol instead. And that is **not** correct, because these few bytes were supposed to be forwarded to anyprot.

**Comment by yrutschle** [comment]
Aaaaaah ok. I think that was an explicit design decision, explained in the comment after read() at the beginning of probe_client_protocol() (probe.c:378 on my tree). Back then I couldn't see anything else than a dropped connection that would result in that, so who cares where it connects as it's going to fail anyway.

Connecting to anyprot in that situation does indeed feel more logical than 'first protocol'. I'll look into it.

**Comment by yrutschle** [comment]
This should do what you expect, can you confirm? (Sorry, I guess I could publish that as a github branch or something, but I'm not familiar with it)
[195.txt](https://github.com/yrutschle/sslh/files/2248647/195.txt)

**Comment by yrutschle** [comment]
Right, get_protocol() is not doing what I thought it was :/ (apparently I've forgotten parts of the logic here...). This probably invalidates the comment I left on your tree, too. I'll try again :/

**Comment by yrutschle** [comment]
Ok, I read through the code to refresh my memory. get_protocol() is arguably not very well named, I might change it. Now I'm thinking your problem is solved if only we returned the *last* protocol on the list, which is usually anyprot (and if it's not, it doesn't really matter where we connect).

**Comment by shadowlmd** [comment]
I wonder why I didn't think of such a simple solution. Sounds good. :)

**Comment by yrutschle** [comment]
Yes, that ought to do it. It's slightly suboptimal because we already iterate over the protocol list just before, and I think we only get to that point if we've iterated through the entire list, but then again that's only triggered when receiving new connections for which no probe work so the performance hit is probably negligeable.

FYI I'm planning to update the entire function taking in account the two issues you raised, but I'm uneasy doing that until I have tests in place for that, so I'm off writing tests first before writing code...
(Obviously I'll have a look at what you've done :) )

**Comment by shadowlmd** [comment]
Yep, that's why I call it an ugly hack. :) Thank you for your work!

**Comment by yrutschle** [comment]
Should be fixed in 677e385fec445987e76da9270a982bdf3d909a92, which is hopefully not too ugly.

**Comment by shadowlmd** [comment]
Thanks a lot, works like a charm! :)
